File: insertsort.js
// i ← 1
// while i < length(A)
//     j ← i
//     while j > 0 and A[j-1] > A[j]
//         swap A[j] and A[j-1]
//         j ← j - 1
//     end while
//     i ← i + 1
// end while

// function insertionSort(array) {
//  let i = 1;
//  while (i < array.length) {
//    let j = 1;
//    while (j > 0 && array[j - 1] > a[j]) {

//    }
//  }
// }
function swap(array, pos1, pos2) {
  let pos1ValRef = array[pos1];
  array[pos1] = array[pos2];
  array[pos2] = pos1ValRef;
}

arr = [12,32,1,52,4,12,23,24,25,8]

// swap(arr, 1, 3)

// console.log(arr)

// function insertionSort(array) {
//   for (let i = 0; i < array.length; i++) {
//     // check if ith position is less than prior positions
//     let j = i - 1;
//     while (j >= 0 && array[i] < array[j]) {
//       // console.log('here')
//       swap(array, i, j);
//       console.log(array)
//       j--;
//       console.log(j);
//     }
//   }
//   console.log(array);
//   return array;
// }

function insertionSort(array) {
  for(let i = 0; i < array.length; i++) {
    let temp = array[i];
    let j = i - 1;
    while (j >= 0 && array[j] > temp) {
      array[j + 1] = array[j];
      j--;
    }
    array[j + 1] = temp;
  }
  return array;
}


insertionSort(arr);

console.log(arr);

File: insertsort.c
#include <stdio.h>

void insertionSort(int arr[], int n)
{
    // iterate through each element in arr
    // start at index 1
    for(int i = 1; i < n; i++) {
        // create temp var to store current iteratations val
        int tmp = arr[i];
        // get the current iterations, previous's index
        int prevI = i - 1;

        // while the previous index is > 0 and
        // its value at previous index is > than currents value
        // swap the values

        while(arr[prevI] > tmp) {
            // swap
            arr[prevI + 1] = arr[prevI];
            // go one step to the left
            prevI--;
        }
        // set final iteration to the current tmp val
        arr[prevI + 1] = tmp;
    }
}

void printArray(int arr[], int n)
{
    for (int i = 0; i < n; i++)
    {
        printf("%d\n", arr[i]);
    }
}

int main(int argc, char *argv[])
{
    int arr[] = {100, 55, 4, 98, 10, 18, 90, 95, 43, 11, 47, 67, 89, 42, 49, 79};
    int n = sizeof(arr) / sizeof(arr[0]);

    insertionSort(arr, n);
    printArray(arr, n);

    return 0;
}